File: src/utils/spatialEngine.js
/**
 * SkyGuard-AI — Nearby Station Spatial Intelligence Engine
 * 
 * Provides:
 * 1. Geodetic distance computation using the Haversine formula.
 * 2. Nearby peer discovery with configurable radius and temporal freshness gating.
 * 3. Robust neighborhood statistics (Median, Median Absolute Deviation [MAD], residuals).
 * 4. Multi-signal fusion combining Station-Adaptive ML with Neighborhood Spatial Consensus
 *    to classify events into NORMAL, LOCALIZED_ANOMALY, or REGIONAL_EVENT.
 */

/**
 * Calculates great-circle distance between two geographic coordinates using Haversine formula.
 * @param {number} lat1 Latitude of point 1 in degrees
 * @param {number} lon1 Longitude of point 1 in degrees
 * @param {number} lat2 Latitude of point 2 in degrees
 * @param {number} lon2 Longitude of point 2 in degrees
 * @returns {number} Distance in kilometers
 */
export function haversineDistance(lat1, lon1, lat2, lon2) {
  if (lat1 === undefined || lon1 === undefined || lat2 === undefined || lon2 === undefined) {
    return Infinity;
  }
  const R = 6371.0; // Earth's mean radius in km
  const toRad = (deg) => (deg * Math.PI) / 180.0;

  const dLat = toRad(lat2 - lat1);
  const dLon = toRad(lon2 - lon1);
  const radLat1 = toRad(lat1);
  const radLat2 = toRad(lat2);

  const a =
    Math.sin(dLat / 2.0) * Math.sin(dLat / 2.0) +
    Math.cos(radLat1) * Math.cos(radLat2) * Math.sin(dLon / 2.0) * Math.sin(dLon / 2.0);
  const c = 2.0 * Math.atan2(Math.sqrt(a), Math.sqrt(1.0 - a));

  return +(R * c).toFixed(2);
}

/**
 * Robust median calculation
 */
function getMedian(arr) {
  if (!arr || arr.length === 0) return 0;
  const sorted = [...arr].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  return sorted.length % 2 !== 0 ? sorted[mid] : (sorted[mid - 1] + sorted[mid]) / 2.0;
}

/**
 * Robust Median Absolute Deviation (MAD)
 */
function getMAD(arr, median) {
  if (!arr || arr.length === 0) return 0;
  const deviations = arr.map((x) => Math.abs(x - median));
  return getMedian(deviations);
}

export class SpatialIntelligenceEngine {
  constructor({ defaultRadiusKm = 50, defaultMaxAgeSeconds = 300 } = {}) {
    this.defaultRadiusKm = defaultRadiusKm;
    this.defaultMaxAgeSeconds = defaultMaxAgeSeconds;
  }

  /**
   * Discovers eligible nearby stations within radius and passing freshness/sanity gates.
   */
  findNearbyStations({ targetStation, stations, radiusKm = this.defaultRadiusKm, maxAgeSeconds = this.defaultMaxAgeSeconds }) {
    if (!targetStation || !stations || stations.length <= 1) {
      return [];
    }

    const now = Date.now();
    const nearby = [];

    for (const st of stations) {
      // Exclude self
      if (st.id === targetStation.id) continue;

      // Coordinate check
      if (st.lat === undefined || st.lon === undefined) continue;

      // Distance calculation
      const dist = haversineDistance(targetStation.lat, targetStation.lon, st.lat, st.lon);
      if (dist > radiusKm) continue;

      // Temporal Freshness Gate: exclude stale dataloggers
      if (st.last_seen) {
        const lastSeenMs = new Date(st.last_seen).getTime();
        if (!isNaN(lastSeenMs)) {
          const ageSeconds = (now - lastSeenMs) / 1000.0;
          if (ageSeconds > maxAgeSeconds) {
            continue; // Stale observation excluded
          }
        }
      }

      // Physical Sanity Gate: Exclude stations reporting hardware error codes (-999, out of bounds)
      const temp = st.sensors?.temperature?.value;
      const hum = st.sensors?.humidity?.value;
      if (temp === undefined || isNaN(temp) || temp < -40 || temp > 65 || hum < 0 || hum > 100) {
        continue; // Corrupt/broken peer excluded from baseline
      }

      nearby.push({
        id: st.id,
        name: st.name,
        region: st.region,
        distance_km: dist,
        lat: st.lat,
        lon: st.lon,
        elevation: st.elevation || 0,
        temp,
        hum: st.sensors?.humidity?.value ?? 60,
        pres: st.sensors?.pressure?.value ?? 1010,
        wind: st.sensors?.wind_speed?.value ?? 10,
        rain: st.sensors?.rainfall?.value ?? 0,
        status: st.status || 'NORMAL',
        ml_model: st.ml_model || null
      });
    }

    // Sort by proximity
    nearby.sort((a, b) => a.distance_km - b.distance_km);
    return nearby;
  }

  /**
   * Computes robust neighborhood statistics and spatial deviation score.
   */
  computeSpatialDeviation({ targetStation, nearbyStations }) {
    if (!nearbyStations || nearbyStations.length === 0) {
      return {
        available: false,
        nearby_count: 0,
        spatial_deviation_score: 0.0,
        spatially_consistent: true,
        reason: "No fresh nearby stations within radius"
      };
    }

    const targetTemp = targetStation.sensors?.temperature?.value ?? 25;
    const targetHum = targetStation.sensors?.humidity?.value ?? 60;
    const targetPres = targetStation.sensors?.pressure?.value ?? 1010;
    const targetRain = targetStation.sensors?.rainfall?.value ?? 0;

    const peerTemps = nearbyStations.map(p => p.temp);
    const peerHums = nearbyStations.map(p => p.hum);
    const peerPress = nearbyStations.map(p => p.pres);
    const peerRains = nearbyStations.map(p => p.rain);

    const medianTemp = getMedian(peerTemps);
    const madTemp = getMAD(peerTemps, medianTemp) || 1.0;
    const residualTemp = Math.abs(targetTemp - medianTemp);

    const medianHum = getMedian(peerHums);
    const residualHum = Math.abs(targetHum - medianHum);

    const medianPres = getMedian(peerPress);
    const residualPres = Math.abs(targetPres - medianPres);

    // Multi-variable normalized deviation
    const tempDev = Math.min(1.0, residualTemp / 5.0); // 5°C delta = 1.0
    const humDev = Math.min(1.0, residualHum / 25.0);  // 25% hum delta = 1.0
    const presDev = Math.min(1.0, residualPres / 6.0); // 6 hPa delta = 1.0

    // Composite spatial deviation score
    const spatialDevScore = +(0.55 * tempDev + 0.25 * humDev + 0.20 * presDev).toFixed(3);
    const spatiallyConsistent = residualTemp <= 3.0 && spatialDevScore < 0.50;

    // Check if nearby stations themselves are exhibiting anomalous weather
    const anomalousPeers = nearbyStations.filter(p => p.ml_model?.is_anomaly || p.status === 'SUSPECT' || p.status === 'EXTREME');
    const peerAnomalyRatio = +(anomalousPeers.length / nearbyStations.length).toFixed(2);

    return {
      available: true,
      nearby_count: nearbyStations.length,
      spatial_deviation_score: spatialDevScore,
      spatially_consistent: spatiallyConsistent,
      neighborhood_median_temp: +medianTemp.toFixed(1),
      neighborhood_mad_temp: +madTemp.toFixed(2),
      residual_temp: +residualTemp.toFixed(1),
      neighborhood_median_hum: +medianHum.toFixed(1),
      residual_hum: +residualHum.toFixed(1),
      peer_anomaly_ratio: peerAnomalyRatio,
      anomalous_peer_count: anomalousPeers.length,
      nearest_peer_id: nearbyStations[0]?.id || null,
      nearest_peer_distance_km: nearbyStations[0]?.distance_km || 0
    };
  }

  /**
   * Fuses Local Station ML with Spatial Neighborhood Consensus to determine final classification.
   */
  fuseAssessment({ physicalQc, localMl, spatialAnalysis }) {
    // 1. Physical Sanity Gate Always Overrides
    if (physicalQc?.fault_risk >= 0.85 && physicalQc?.reason_codes?.includes("RANGE_FAIL")) {
      return {
        classification: "PHYSICAL_SENSOR_FAILURE",
        badge_class: "badge-critical",
        confidence: "VERY_HIGH",
        interpretation: "Raw measurements breached universal physical plausibility bounds. Hardware failure suspected."
      };
    }

    // 2. Cold-start or Isolated Station with Zero Neighbors
    if (!spatialAnalysis?.available || spatialAnalysis.nearby_count === 0) {
      if (localMl?.is_anomaly) {
        return {
          classification: "LOCAL_ANOMALY_UNVERIFIED",
          badge_class: "badge-suspect",
          confidence: "MEDIUM",
          interpretation: "Station-specific model flagged anomaly; no nearby peers within radius for spatial verification."
        };
      }
      return {
        classification: "NORMAL",
        badge_class: "badge-normal",
        confidence: "HIGH",
        interpretation: "Observation nominal according to local station baseline (operating independently)."
      };
    }

    const isLocalAnomaly = !!localMl?.is_anomaly;
    const isSpatiallyConsistent = spatialAnalysis.spatially_consistent;
    const peerAnomalyRatio = spatialAnalysis.peer_anomaly_ratio || 0.0;

    // 3. Both Local Model and Spatial Neighbors Agree: NORMAL
    if (!isLocalAnomaly && isSpatiallyConsistent) {
      return {
        classification: "NORMAL",
        badge_class: "badge-normal",
        confidence: "HIGH",
        interpretation: "Readings consistent with both station historical baseline and nearby peer consensus."
      };
    }

    // 4. Local Model Flags Anomaly BUT Nearby Stations Experience Similar Conditions: REGIONAL WEATHER EVENT
    if (isLocalAnomaly && (isSpatiallyConsistent || peerAnomalyRatio >= 0.4)) {
      return {
        classification: "REGIONAL_EVENT",
        badge_class: "badge-extreme",
        confidence: "HIGH",
        interpretation: `Atmospheric excursion detected locally but corroborated by ${spatialAnalysis.nearby_count} nearby stations within radius. Consistent with regional weather event / storm.`
      };
    }

    // 5. Local Model Flags Anomaly AND Nearby Stations Are Normal: LOCALIZED ANOMALY (SENSOR DEFECT)
    if (isLocalAnomaly && !isSpatiallyConsistent) {
      return {
        classification: "LOCALIZED_ANOMALY",
        badge_class: "badge-critical",
        confidence: "HIGH",
        interpretation: `Local reading deviates ${spatialAnalysis.residual_temp}°C from peer median (${spatialAnalysis.neighborhood_median_temp}°C across ${spatialAnalysis.nearby_count} nearby stations). Likely localized sensor drift or hardware malfunction.`
      };
    }

    // 6. Local Model Normal BUT High Spatial Divergence (Microclimate gradient)
    return {
      classification: "MICROCLIMATE_GRADIENT",
      badge_class: "badge-suspect",
      confidence: "MODERATE",
      interpretation: "Station reading is normal for its own historical distribution, but diverges from nearby peers (terrain/elevation gradient)."
    };
  }

  /**
   * End-to-end neighborhood analysis for a given station.
   */
  analyzeStation({ targetStation, stations, radiusKm = this.defaultRadiusKm, maxAgeSeconds = this.defaultMaxAgeSeconds, localMl, physicalQc }) {
    const nearby = this.findNearbyStations({ targetStation, stations, radiusKm, maxAgeSeconds });
    const spatialDev = this.computeSpatialDeviation({ targetStation, nearbyStations: nearby });
    const assessment = this.fuseAssessment({ physicalQc, localMl, spatialAnalysis: spatialDev });

    return {
      search_radius_km: radiusKm,
      nearby_stations: nearby,
      spatial_analysis: spatialDev,
      final_assessment: assessment
    };
  }
}

export const spatialEngine = new SpatialIntelligenceEngine();

File: ml/spatial_engine.py
#!/usr/bin/env python3
"""
SkyGuard-AI — Spatial Intelligence Engine (Python Implementation)
Provides:
1. Geodetic distance using the Haversine formula.
2. Temporal freshness and physical sanity filtering of nearby peers.
3. Robust neighborhood statistics (Median, MAD, multi-variable residuals).
4. Dual-Track Anomaly Fusion combining Station-Adaptive ML with Spatial Consensus.
"""

import math
from typing import List, Dict, Any, Optional

def haversine_distance(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """Calculates great-circle distance between two geographic coordinates in kilometers."""
    if None in (lat1, lon1, lat2, lon2):
        return float("inf")
    R = 6371.0  # Earth's mean radius in km

    d_lat = math.radians(lat2 - lat1)
    d_lon = math.radians(lon2 - lon1)
    rad_lat1 = math.radians(lat1)
    rad_lat2 = math.radians(lat2)

    a = (math.sin(d_lat / 2.0) ** 2) + math.cos(rad_lat1) * math.cos(rad_lat2) * (math.sin(d_lon / 2.0) ** 2)
    c = 2.0 * math.atan2(math.sqrt(a), math.sqrt(1.0 - a))
    return round(R * c, 2)


def get_median(arr: List[float]) -> float:
    if not arr:
        return 0.0
    s = sorted(arr)
    n = len(s)
    mid = n // 2
    return s[mid] if n % 2 != 0 else (s[mid - 1] + s[mid]) / 2.0


def get_mad(arr: List[float], median: float) -> float:
    if not arr:
        return 0.0
    devs = [abs(x - median) for x in arr]
    return get_median(devs)


class SpatialIntelligenceEngine:
    def __init__(self, default_radius_km: float = 50.0, default_max_age_seconds: float = 300.0):
        self.default_radius_km = default_radius_km
        self.default_max_age_seconds = default_max_age_seconds

    def find_nearby_stations(
        self,
        target_station: Dict[str, Any],
        stations: List[Dict[str, Any]],
        radius_km: Optional[float] = None,
        max_age_seconds: Optional[float] = None,
        current_timestamp: Optional[float] = None,
    ) -> List[Dict[str, Any]]:
        radius = radius_km if radius_km is not None else self.default_radius_km
        max_age = max_age_seconds if max_age_seconds is not None else self.default_max_age_seconds
        now = current_timestamp if current_timestamp is not None else 1000000.0

        t_lat = target_station.get("lat")
        t_lon = target_station.get("lon")
        t_id = target_station.get("id")

        if t_lat is None or t_lon is None:
            return []

        nearby = []
        for st in stations:
            if st.get("id") == t_id:
                continue

            s_lat = st.get("lat")
            s_lon = st.get("lon")
            if s_lat is None or s_lon is None:
                continue

            dist = haversine_distance(t_lat, t_lon, s_lat, s_lon)
            if dist > radius:
                continue

            # Temporal Freshness Check
            last_seen = st.get("last_seen_epoch", now)
            age = now - last_seen
            if age > max_age:
                continue  # Exclude stale readings

            # Physical Sanity Check
            readings = st.get("readings", st.get("sensors", {}))
            temp = readings.get("temp", readings.get("temperature", None))
            hum = readings.get("hum", readings.get("humidity", None))

            if temp is None or temp < -40.0 or temp > 65.0:
                continue  # Exclude corrupted/invalid hardware readings
            if hum is not None and (hum < 0.0 or hum > 100.0):
                continue

            nearby.append({
                "id": st.get("id"),
                "name": st.get("name", st.get("id")),
                "distance_km": dist,
                "lat": s_lat,
                "lon": s_lon,
                "elevation": st.get("elevation", 0),
                "temp": float(temp),
                "hum": float(hum) if hum is not None else 60.0,
                "pres": float(readings.get("pres", readings.get("pressure", 1010.0))),
                "status": st.get("status", "NORMAL"),
                "is_anomaly": st.get("is_anomaly", False),
            })

        nearby.sort(key=lambda x: x["distance_km"])
        return nearby

    def compute_spatial_deviation(
        self,
        target_station: Dict[str, Any],
        nearby_stations: List[Dict[str, Any]]
    ) -> Dict[str, Any]:
        if not nearby_stations:
            return {
                "available": False,
                "nearby_count": 0,
                "spatial_deviation_score": 0.0,
                "spatially_consistent": True,
                "reason": "No fresh nearby stations within radius"
            }

        readings = target_station.get("readings", target_station.get("sensors", {}))
        target_temp = float(readings.get("temp", readings.get("temperature", 25.0)))
        target_hum = float(readings.get("hum", readings.get("humidity", 60.0)))
        target_pres = float(readings.get("pres", readings.get("pressure", 1010.0)))

        peer_temps = [float(p.get("temp", p.get("readings", {}).get("temp", 25.0))) for p in nearby_stations]
        peer_hums = [float(p.get("hum", p.get("readings", {}).get("hum", 60.0))) for p in nearby_stations]
        peer_press = [float(p.get("pres", p.get("readings", {}).get("pres", 1010.0))) for p in nearby_stations]

        med_temp = get_median(peer_temps)
        mad_temp = get_mad(peer_temps, med_temp) or 1.0
        res_temp = abs(target_temp - med_temp)

        med_hum = get_median(peer_hums)
        res_hum = abs(target_hum - med_hum)

        med_pres = get_median(peer_press)
        res_pres = abs(target_pres - med_pres)

        temp_dev = min(1.0, res_temp / 5.0)
        hum_dev = min(1.0, res_hum / 25.0)
        pres_dev = min(1.0, res_pres / 6.0)

        spatial_score = round(0.55 * temp_dev + 0.25 * hum_dev + 0.20 * pres_dev, 3)
        spatially_consistent = res_temp <= 3.0 and spatial_score < 0.50

        anomalous_peers = [p for p in nearby_stations if p.get("is_anomaly") or p.get("status") in ("SUSPECT", "EXTREME")]
        peer_anomaly_ratio = round(len(anomalous_peers) / len(nearby_stations), 2)

        return {
            "available": True,
            "nearby_count": len(nearby_stations),
            "spatial_deviation_score": spatial_score,
            "spatially_consistent": spatially_consistent,
            "neighborhood_median_temp": round(med_temp, 1),
            "neighborhood_mad_temp": round(mad_temp, 2),
            "residual_temp": round(res_temp, 1),
            "neighborhood_median_hum": round(med_hum, 1),
            "residual_hum": round(res_hum, 1),
            "peer_anomaly_ratio": peer_anomaly_ratio,
            "anomalous_peer_count": len(anomalous_peers),
            "nearest_peer_id": nearby_stations[0].get("id") if nearby_stations else None,
            "nearest_peer_distance_km": nearby_stations[0].get("distance_km", 0.0) if nearby_stations else 0.0,
        }

    def fuse_assessment(
        self,
        physical_qc: Optional[Dict[str, Any]],
        local_ml: Optional[Dict[str, Any]],
        spatial_analysis: Dict[str, Any]
    ) -> Dict[str, Any]:
        # 1. Physical QC Gate Always Overrides
        if physical_qc and physical_qc.get("fault_risk", 0.0) >= 0.85 and "RANGE_FAIL" in physical_qc.get("reason_codes", []):
            return {
                "classification": "PHYSICAL_SENSOR_FAILURE",
                "confidence": "VERY_HIGH",
                "interpretation": "Raw sensor reading breached universal physical plausibility envelope."
            }

        # 2. No Peers Available (Cold Start / Remote Station)
        if not spatial_analysis.get("available") or spatial_analysis.get("nearby_count", 0) == 0:
            if local_ml and local_ml.get("is_anomaly"):
                return {
                    "classification": "LOCAL_ANOMALY_UNVERIFIED",
                    "confidence": "MEDIUM",
                    "interpretation": "Local ML flagged anomaly; no nearby peers within radius for spatial corroboration."
                }
            return {
                "classification": "NORMAL",
                "confidence": "HIGH",
                "interpretation": "Nominal according to station baseline; operating without nearby spatial peers."
            }

        is_local_anomaly = bool(local_ml and local_ml.get("is_anomaly"))
        is_spatially_consistent = spatial_analysis.get("spatially_consistent", True)
        peer_anomaly_ratio = spatial_analysis.get("peer_anomaly_ratio", 0.0)

        # 3. Both Local Model and Spatial Peers Agree Nominal
        if not is_local_anomaly and is_spatially_consistent:
            return {
                "classification": "NORMAL",
                "confidence": "HIGH",
                "interpretation": "Readings consistent with both station historical baseline and nearby peer consensus."
            }

        # 4. Local Model Flags Anomaly BUT Nearby Stations Also Experience Similar Extreme Weather
        if is_local_anomaly and (is_spatially_consistent or peer_anomaly_ratio >= 0.4):
            return {
                "classification": "REGIONAL_EVENT",
                "confidence": "HIGH",
                "interpretation": f"Atmospheric excursion detected locally but corroborated by nearby peers within {spatial_analysis.get('nearby_count')} stations. Consistent with regional storm or front."
            }

        # 5. Local Model Flags Anomaly AND Nearby Stations Are Normal (Sensor Defect)
        if is_local_anomaly and not is_spatially_consistent:
            return {
                "classification": "LOCALIZED_ANOMALY",
                "confidence": "HIGH",
                "interpretation": f"Local reading deviates {spatial_analysis.get('residual_temp')}°C from peer median ({spatial_analysis.get('neighborhood_median_temp')}°C across {spatial_analysis.get('nearby_count')} nearby stations). Likely localized sensor drift or defect."
            }

        # 6. Local Model Normal BUT High Spatial Divergence
        return {
            "classification": "MICROCLIMATE_GRADIENT",
            "confidence": "MODERATE",
            "interpretation": "Station reading is normal for its own local history, but diverges from nearby peers (terrain or elevation gradient)."
        }
